src/utils/encoding/mod.rs

```diff
@@ -79,15 +83,123 @@ pub fn prover_id_from_u64(id: u64) -> ProverId {
     prover_id
 }
 
+// Unsigned VarInt (Uvi) Bytes
+pub struct UviBytes {
+    // cache for varint frame size
+    len: Option<usize>,
+    // content size limit, defaults to 128MiB
+    limit: usize,
+}
+
+impl UviBytes {
+    pub fn new_with_limit(limit: usize) -> Self {
+        UviBytes { len: None, limit }
+    }
+}
+
+impl Default for UviBytes {
+    fn default() -> Self {
+        UviBytes::new_with_limit(128 * 1024 * 1024)
+    }
+}
+
+impl Decoder for UviBytes {
+    type Item = BytesMut;
+    type Error = io::Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        if self.len.is_none() {
```

Comment:
Why not having this as an `else` branch to the next condition statement?

Reply:
Nope. :(

Reply:
Because it has to deal with partial decoding. There are three states: Insufficient data to parse the frame length, invalid frame length, got header length (and doing incremental reads of the frame body).